### DISCUSSION TITLE
Using adjoint solver for varying holes radius in a photonic crystal.

### DISCUSSION THREAD

**Discussion by mariusCZ** [NanoComp/meep#2391] [Q&A] [answered]
Hi everyone, I have a photonic crystal (PhC) structure formed from triangular lattice of holes in a dielectric where there is a cavity and a line defect. I need to optimize the coupling between the cavity and the waveguide, meaning I need to find the parameters which would lead to highest transmission in the waveguide. So I was wondering, would it be possible to use the adjoint solver to try and optimize hole radius for transmission?

I have read other discussions here and it seems that the solutions are with significant caveats. So I guess the better question is if there is any progress on non-density based optimization using the adjoint solver, which would allow for solving this problem? If not, this may not be the best place to ask, but would it be possible to use scipy together with autograd to try and do this? Thank you!

**Comment by smartalecH** [comment, marked as answer]
So it sounds like you've already read the various discussions that talk about our hybrid density-based TO approach, which allows you to use your density variables as a level set. This is important, because a level set is needed in order to do shape optimization. 

To answer your question, there hasn't been much progress. But we are starting to ramp up development here again. 

To answer your other question, simply plopping in scipy and autograd isn't going to solve your problem (we already use autograd?) To be clear, the issue isn't with gradients. Meep computes gradients well through the whole pipeline, and relies on autograd when necessary. The issue is with the parameterization and how it interacts with the PDE.

So writing your own implementation would still require you to figure out how to produce a level set for your PhC holes. 

I'll say that I think your problem is better suited for density-based methods anyway. You'll get much better coupling efficiency between the modes. In fact, using a full Maxwell adjoint method seems overkill if all you're optimizing is a hole radius. 

I'll also add that an FDTD method probably isn't a suitable choice when part of your problem involves a high Q resonance... it's going to take forever to resolve the dft fields. Unless you can incorporate a Pade approximate.

**Reply by mariusCZ** [reply]
As much as I would like to use density-based methods, I am afraid manufacturing them isn't quite easy and I must take that in to account. 

Perhaps in this case something like a genetic algorithm is better suited? Also, what other options would I even have besides FDTD for looking at transmission?

**Reply by smartalecH** [reply]
> I am afraid manufacturing them isn't quite easy and I must take that in to account.

Robust TO is a versatile technique to combat this. I've designed and fabricated several different devices and systems on complicated, commercial foundry platforms this way:

https://smartech.gatech.edu/handle/1853/67338

Keep in mind that when designing highly resonant devices, you're always going to produce a device which is very sensitive to manufacturing variability, regardless of your parameterization or optimization approach. 

If you're more concerned about what kind of feature sizes/types you can allow (because the material platform is hard to work with), there are plenty of techniques for that too:

https://doi.org/10.1364/OE.431188

https://arxiv.org/abs/2212.09302

> Perhaps in this case something like a genetic algorithm is better suited?

Sure, anytime you only have 1 or 2 degrees of freedom, you can really try whatever algorithm you want. Gradient-based algorithms really shine when the number of DOF increases, though. 

> what other options would I even have besides FDTD for looking at transmission?

Anything in the frequency domain. FDFD, FEM, RCWA, etc. @mochen does a lot of inverse design of resonant devices. He could add more on that topic than I could.

**Reply by oskooi** [reply]
In case it may be helpful, we recently added a new tutorial for the adjoint solver which demonstrates incorporating minimum feature sizes in the optimization based on the method that @smartalecH referenced:

https://meep.readthedocs.io/en/latest/Python_Tutorials/Adjoint_Solver/#broadband-waveguide-mode-converter-with-minimum-feature-size

This tutorial reviews several important aspects of setting up these types of optimization problems so it deserves careful study.

**Reply by mariusCZ** [reply]
> Keep in mind that when designing highly resonant devices, you're always going to produce a device which is very sensitive to manufacturing variability, regardless of your parameterization or optimization approach.

I am aware that the devices will not be exactly robust, but that is not the main concern. We are limited to in house device manufacturing and a lot of inverse design devices seem to island of bulk material. In the case of our manufacturing capabilities, we can't have unconnected semiconductor material, as otherwise it just floats away for reason beyond my knowledge.

> Anything in the frequency domain. FDFD, FEM, RCWA, etc. @mochen does a lot of inverse design of resonant devices. He could add more on that topic than I could.

I realise looking at resonant devices in frequency domain is good in terms of looking at Q factor, mode profiles and such, however, correct me if I am wrong, frequency domain simulations will not show transmission.

Anyways, thank you very much for all your help and I really appreciate all the work you are putting in to meep, it's an invaluable tool.

**Reply by smartalecH** [reply]
> In the case of our manufacturing capabilities, we can't have unconnected semiconductor material,

Sounds like a perfect use case of our connectivity constraints 

> however, correct me if I am wrong, frequency domain simulations will not show transmission.

You can indeed compute transmission using frequency-domain codes. Technically, meep computes it's transmission in the frequency domain using post processing.
